### ISSUE TITLE
Link drivers statically

### ISSUE THREAD

**Issue by sblendorio** [cc65/cc65#1399]
I read this documentation about statically link drivers using **co65**:
https://cc65.github.io/doc/co65.html

...but I didn't realize how to link statically, for example, **plus4-stdser.ser** for Plus4's serial port. For instance, there is a reference to **tgi_install** function, which is specific to "TGI", so I don't know which is the corresponding for plus4-stdser.

Referring to @mrdudz's comment here: https://github.com/cc65/cc65/issues/1398#issuecomment-776711735

How could I do it?

Thanks in advance,
Francesco

**Comment by oliverschmidt** [comment]
https://github.com/cc65/cc65/blob/master/include/serial.h#L139

https://github.com/cc65/cc65/blob/master/include/plus4.h#L59

**Comment by groessler** [comment]
You don't need to fiddle with co65. Just link the driver.

Here's an example (which doesn't work anymore since cc65 lost the `systime()` function, but it should show how to include the serial driver.

[serdump.zip](https://github.com/cc65/cc65/files/5965912/serdump.zip)

**Comment by polluks** [comment]
@groessler BTW should co65.sgml be tagged "obsolete"?

**Comment by groessler** [comment]
Hmm. Reading the documentation it appears that it's main purpose was to convert dynamic drivers to static ones. But some years ago already the cc65 build system was changed to create static driver object files alongside with the dynamic drivers.

I there's no other use for co65 I tend to tag it 'obsolete'.

@oliverschmidt, what do you think?

**Comment by oliverschmidt** [comment]
I'd be absolutely fine with - and would in fact advocate for - removing co65 altogether right now.

**Comment by oliverschmidt** [comment]
@sblendorio: Sorry if my https://github.com/cc65/cc65/issues/1399#issuecomment-777472958 above should have been too brief. Just forget about anything else and simply use the two links I mentioned. That's all to it. Your driver is loaded statically with just placing that call in your program.

**Comment by polluks** [comment]
@oliverschmidt co65 should be kept for old projects...

**Comment by mrdudz** [comment]
> co65 should be kept for old projects...

What would be the usecase?

**Comment by polluks** [comment]
Old makefiles are still working.

**Comment by mrdudz** [comment]
Fixing that one line in the makefile will be the least of your problems with making any non trivial old project work with recent cc65

**Comment by polluks** [comment]
Indeed, however we are pretty backward compatible since 20 years :-)

**Comment by oliverschmidt** [comment]
How about this: Remove it completely from the documentation, but keep the actual code/build/binary.

**Comment by sblendorio** [comment]
@oliverschmidt I see this comment:

    unsigned char __fastcall__ ser_install (void* driver);
    /* Install an already loaded driver. Return an error code. */

This means that before calling **ser_install** I should call **ser_load_driver**?

**Comment by oliverschmidt** [comment]
No. As I wrote: Just that single call.

**Comment by sblendorio** [comment]
Ok. Should the file **plus4-stdser.ser** be in the same directory of the **.c** source file?

**Comment by mrdudz** [comment]
> Indeed, however we are pretty backward compatible since 20 years :-)

I just recently assisted someone with updating a project he didnt touch for a couple years, and it involved more than some trivial changes - so no, we are not. And breaking things mercylessly was always the spirit of cc65 development too. *Especially* with linker related things.

> Remove it completely from the documentation, but keep the actual code/build/binary.

I'd just point to the old binary and have a note that its depcrecated and no more needed. Letting it dangling around with no documentation creates more problems than it solves, IMHO.

**Comment by oliverschmidt** [comment]
> Ok. Should the file plus4-stdser.ser be in the same directory of the .c source file?

No, forget about that file. As I wrote: Just that single call.

**Comment by oliverschmidt** [comment]
> And breaking things mercylessly was always the spirit of cc65 development too. Especially with linker related things.

I confess that this is my position. The momentum behind the cc65 development is way to small to be invested into backward compatibility (thoughts). That's in line with my proposal here to remove either everything or just the docu.

> I'd just point to the old binary and have a note that its depcrecated and no more needed.

I'm willing to merge a PR doing so.

**Comment by polluks** [comment]
> @oliverschmidt I see this comment:
> 
> ```
> unsigned char __fastcall__ ser_install (void* driver);
> /* Install an already loaded driver. Return an error code. */
> ```
> 
> This means that before calling **ser_install** I should call **ser_load_driver**?

A static driver is always "already loaded" because it's part of the prg :-)

**Comment by groessler** [comment]
Did we come to an agreement about what to do with `co65`?

My opinion is "keep it", it might be useful in some cases. But remove all reference to it when talking about static drivers.

**Comment by oliverschmidt** [comment]
>> @mrdudz: I'd just point to the old binary and have a note that its depcrecated and no more needed.

> @oliverschmidt: I'm willing to merge a PR doing so.

I'd classify the above as an agreement - as nobody opposed to it:

> My opinion is "keep it", it might be useful in some cases. But remove all references to it when talking about static drivers.

As far as I understand your opinion  aligns pretty well with the "agreement" above :-)

**Comment by mrdudz** [comment]
closing. this is a non issue by now
